### ISSUE TITLE
Metabase does not install new version of plugins when upgrading

### ISSUE THREAD

**Issue by dacort** [metabase/metabase#11699]
**Describe the bug**
Metabase 0.34 included several time-related changes, but when upgrading Metabase it does not upgrade the plugin jars in the `plugins` directory resulting in errors when creating new databases. This will also cause queries on existing databases to break.

**Logs**
When trying to add a database for anything like Oracle, SQLite, or SQLServer, the first attempt at hitting "Save" results in this error in the browser:
```
Syntax error compiling at (metabase/driver/oracle.clj:1:1).
```

Subsequent attempts to save result in this error:
```
No method in multimethod 'connection-details->spec' for dispatch value: :oracle
```

Server logs for the first "Save" request:
```

01-13 08:42:00 INFO driver.impl :: Initializing driver :oracle...
01-13 08:42:00 INFO plugins.classloader :: Added URL file:/private/tmp/metabase-upgrade-plugins/plugins/oracle.metabase-driver.jar to classpath
01-13 08:42:00 DEBUG plugins.init-steps :: Loading plugin namespace metabase.driver.oracle...
01-13 08:42:00 ERROR driver.util :: Database connection error
Syntax error compiling at (metabase/driver/oracle.clj:1:1).
	at clojure.lang.Compiler.load(Compiler.java:7648)
	at clojure.lang.RT.loadResourceScript(RT.java:381)
	at clojure.lang.RT.loadResourceScript(RT.java:372)
	at clojure.lang.RT.load(RT.java:459)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.plugins.classloader$require.invokeStatic(classloader.clj:101)
	at metabase.plugins.classloader$require.doInvoke(classloader.clj:89)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at metabase.plugins.init_steps$fn__67727.invokeStatic(init_steps.clj:22)
	at metabase.plugins.init_steps$fn__67727.invoke(init_steps.clj:20)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.plugins.init_steps$do_init_steps_BANG_.invokeStatic(init_steps.clj:32)
	at metabase.plugins.init_steps$do_init_steps_BANG_.invoke(init_steps.clj:27)
	at metabase.plugins.lazy_loaded_driver$make_initialize_BANG_$fn__67766.invoke(lazy_loaded_driver.clj:58)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.driver.impl$initialize_if_needed_BANG_.invokeStatic(impl.clj:179)
	at metabase.driver.impl$initialize_if_needed_BANG_.invoke(impl.clj:160)
	at metabase.driver$the_initialized_driver.invokeStatic(driver.clj:146)
	at metabase.driver$the_initialized_driver.invoke(driver.clj:142)
	at metabase.driver$dispatch_on_initialized_driver.invokeStatic(driver.clj:154)
	at metabase.driver$dispatch_on_initialized_driver.doInvoke(driver.clj:149)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at clojure.lang.MultiFn.invoke(MultiFn.java:233)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__21110.invoke(util.clj:32)
	at metabase.util$do_with_timeout$fn__13131.invoke(util.clj:302)
	at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.FileNotFoundException: Could not locate metabase/util/date__init.class, metabase/util/date.clj or metabase/util/date.cljc on classpath.
	at clojure.lang.RT.load(RT.java:462)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5989)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.invoke(RestFn.java:703)
	at metabase.driver.oracle$eval192$loading__6721__auto____193.invoke(oracle.clj:1)
	at metabase.driver.oracle$eval192.invokeStatic(oracle.clj:1)
	at metabase.driver.oracle$eval192.invoke(oracle.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7177)
	at clojure.lang.Compiler.eval(Compiler.java:7166)
	at clojure.lang.Compiler.load(Compiler.java:7636)
	... 50 more
01-13 08:42:00 DEBUG middleware.log :: POST /api/database 400 50.8 ms (0 DB calls)
{:valid false, :dbname "Syntax error compiling at (metabase/driver/oracle.clj:1:1).", :message "Syntax error compiling at (metabase/driver/oracle.clj:1:1)."}
```

**To Reproduce**
1. Download an older version of Metabase and run it
2. Download a newer jar and run that in the same directory
3. Note that none of the dates on the jars in `plugins/` change
4. Try to add a database for Oracle, SQLite, SQLServer, etc
5. See the error message mentioned above

**Expected behavior**
Running a new version of Metabase should upgrade the jar files in the `plugins` directory

**Screenshots**
n/a

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.117 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.2+9",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "11.0.2",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.2+9",
    "os.name": "Mac OS X",
    "os.version": "10.14.6",
    "user.language": "en",
    "user.timezone": "America/Los_Angeles"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "oracle"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "run-mode": "prod",
    "version": {
      "date": "2019-12-19",
      "tag": "v0.34.0",
      "branch": "release-0.34.x",
      "hash": "1335215"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Critical - breaks basic functionality and will also break existing database connections. 

**Additional context**
n/a

**Comment by dacort** [comment]
We _do_ check to see if the destination file is older than the source file in [`copy-file!`](https://github.com/metabase/metabase/blob/412c115374364e4a54b49c0ecd52d68903b2e520/src/metabase/util/files.clj#L76), so it's possible this may be a timing issue.

For example, if somebody downloaded and installed an earlier release _after_ December 19th (when 0.34 was released) and then tried to upgrade to 0.34, this issue would occur. But in most normal circumstances, the plugin jars should be overwritten with newer versions.

**Comment by dacort** [comment]
Confirmed that Metabase will overwrite the plugin jar if the modified attribute is older than the one in the uberjar. Perhaps we include the `COPY_ATTRIBUTES` option when doing the file copy?

- Run Metabase 0.33.6
- Note that last modification date of files is now
    ````
    › stat -x plugins/sqlite.metabase-driver.jar
    File: "plugins/sqlite.metabase-driver.jar"
    Size: 7115610      FileType: Regular File
    Mode: (0644/-rw-r--r--)         Uid: (  [redacted])  Gid: (    [redacted])
    Device: 1,4   Inode: 33982694    Links: 1
    Access: Mon Jan 13 09:22:24 2020
    Modify: Mon Jan 13 09:21:44 2020
    Change: Mon Jan 13 09:21:44 2020
    ```
- Update the "date modified" attribute to be before Dec 19
    ```shell
    touch -mt 201912120000 plugins/sqlite.metabase-driver.jar
    ```
- Run Metabase 0.34.0
- Note that we extract the sqlite driver
    ```
    Extract file /modules/sqlite.metabase-driver.jar -> /private/tmp/metabase-upgrade-plugins/plugins/sqlite.metabase-driver.jar took 39.7 ms
    ```

**Comment by camsaul** [comment]
@dacort I'll take a look at this

**Comment by camsaul** [comment]
@dacort I was thinking about more robust ways to fix this, interested in your opinion. I think ideally we actually want to overwrite the plugins any time the version being ran is different than the version that extracted the existing plugins. That way, if you downgrade, you're not stuck with newer plugins that potentially are incompatible.

I thought maybe using [`UserDefinedFileAttributeView`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/attribute/UserDefinedFileAttributeView.html) could be a good option, storing the MB version associated with each plugin, but it doesn't look like it's supported on all platforms (at least it's not working on macOS)

I think maybe setting the last modified time to build date would be a good solution? e.g.

```clj
(metabase.util.date-2/parse (:date metabase.config/mb-version-info))
```

And then if the build date for the current build isn't the same as the last modified timestamp for the plugins we can overwrite them. Does that make sense?

**Comment by camsaul** [comment]
Maybe it makes sense to copy the last modified attribute like you suggested and then just do a simple equality comparison for the last modified timestamps instead of seeing the one potentially being copied is newer

e.g.

```clj
(defn- last-modified-timestamp
  "Last modified timestamp for a file at `path`."
  ^java.time.Instant [^Path path]
  (.toInstant (Files/getLastModifiedTime path (u/varargs LinkOption))))

(defn- copy-file! [^Path source, ^Path dest]
  (when (or (not (exists? dest))
            (not= (last-modified-timestamp source) (last-modified-timestamp dest)))
    (u/profile (trs "Extract file {0} -> {1}" source dest)
      (Files/copy source dest (u/varargs CopyOption [StandardCopyOption/REPLACE_EXISTING
                                                     StandardCopyOption/COPY_ATTRIBUTES])))))
```

**Comment by dacort** [comment]
@camsaul Ah, yea, the downgrade scenario is interesting and makes it more tricky. I would say that using `COPY_ATTRIBUTES` and doing a simple equality comparison sounds like a good approach assuming that it's supported on all required platforms. But even in the case that it's not, overwriting the plugins on start wouldn't be terrible.

**Comment by camsaul** [comment]
Fixed by #11701
